consensus/state_processing/src/common/get_attestation_participation.rs

```diff
@@ -0,0 +1,46 @@
+use crate::per_block_processing::errors::BlockProcessingError as Error;
+use integer_sqrt::IntegerSquareRoot;
+use safe_arith::SafeArith;
+use smallvec::SmallVec;
+use types::consts::altair::{
+    NUM_FLAG_INDICES, TIMELY_HEAD_FLAG_INDEX, TIMELY_SOURCE_FLAG_INDEX, TIMELY_TARGET_FLAG_INDEX,
+};
+use types::{AttestationData, BeaconState, ChainSpec, EthSpec};
+
+/// Get the participation flags for a **valid** attestation.
+///
+/// You must have called `verify_attestation_for_block_inclusion` or similar before
+/// calling this function, in order to ensure that the attestation's source is correct.
+///
+/// This function is extracted from `process_attestation`
+pub fn get_attestation_participation<T: EthSpec>(
+    data: &AttestationData,
+    state: &BeaconState<T>,
+    spec: &ChainSpec,
+) -> Result<SmallVec<[u32; NUM_FLAG_INDICES]>, Error> {
+    // Matching roots.
+    // Source match is checked by `verify_attestation_for_block_inclusion`.
+    let is_matching_head = data.beacon_block_root == *state.get_block_root(data.slot)?;
+    let is_matching_source = true;
```

Comment:
Why don't we check the source here? I haven't been able to figure it out yet, perhaps it's something do with the op pool?

If it's just for efficiency, perhaps the gains aren't worth the risk of something "falling through the cracks"?

Reply:
Yeah, you're right that it doesn't hurt to have that check in there. This function existed as a sort of weird intermediate that I had to yank out of attestation processing, but it's now been refactored in the spec into a real entity: https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.6/specs/altair/beacon-chain.md#get_attestation_participation_flag_indices

I'll fix it on this branch, and then carry that over when we do the update

Reply:
Resolved in https://github.com/sigp/lighthouse/pull/2279/commits/6415962290a54c00289cb526aefc5a304e9e94e4